### PR TITLE
DDS tools/plugin/control-ui update and misc.

### PR DESCRIPTION
**Pull request by rbx**
Builds upon changes proposed in #135, supersedes it.
Adds functionality for future developments of `fair::mq::sdk`:
- `FairMQDevice::TransitionTo` (synchronous).
- Extract States and Transitions declarations to its own header, use in plugins and device.
- Extracts state queue functionality into its own class, use in device and plugins.
---
Checklist:

* [x] Rebased against `dev` branch
* [x] My name is in the resp. CONTRIBUTORS/AUTHORS file
* [x] Followed [the seven rules of great commit messages](https://chris.beams.io/posts/git-commit/#seven-rules)

### CODE REVIEW

File: fairmq/plugins/DDS/runDDSCommandUI.cxx [hunk 1]
```diff
@@ -8,146 +8,269 @@
...
 
-        bpo::options_description options("fairmq-dds-command-ui options");
         options.add_options()
-            ("session,s", bpo::value<string> (&sessionID)->required(), "DDS Session ID")
-            ("command,c", bpo::value<char>   (&command)->default_value(' '), "Command character")
-            ("path,p",    bpo::value<string> (&topologyPath)->default_value(""), "DDS Topology path to send command to")
+            ("command,c",        bpo::value<string> (&command)->default_value(""), "Command character")
+            ("path,p",           bpo::value<string> (&topologyPath)->default_value(""), "DDS Topology path to send command to (empty - send to all tasks)")
+            ("wait-for-state,w", bpo::value<string> (&targetState)->default_value(""), "Wait until targeted FairMQ devices reach the given state")
+            ("timeout,t",        bpo::value<unsigned int> (&timeout)->default_value(0), "Timeout in milliseconds when waiting for a device state (0 - wait infinitely)")
```
[the remark this answers is not in the document]
Comment (rbx) [reply fairmq/plugins/DDS/runDDSCommandUI.cxx]: False positive.

File: fairmq/plugins/DDS/runDDSCommandUI.cxx [hunk 2]
```diff
@@ -8,146 +8,269 @@
...
+        } else {
+            cin >> c;
+            command = c;
+        }
+    }
+}
+
+struct WaitMode
+{
+    WaitMode(const string& targetState)
```
[the remark this answers is not in the document]
Comment (rbx) [reply fairmq/plugins/DDS/runDDSCommandUI.cxx]: False positive.

File: fairmq/plugins/DDS/runDDSCommandUI.cxx [hunk 3]
```diff
@@ -8,146 +8,269 @@
...
+            cout << "> end" << endl;
+            ddsCustomCmd.send("END", topologyPath);
+        } else if (command == "q!") {
+            cout << "> shutdown" << endl;
+            ddsCustomCmd.send("SHUTDOWN", topologyPath);
+        } else if (command == "r!") {
+            cout << "> startup" << endl;
+            ddsCustomCmd.send("STARTUP", topologyPath);
+        } else {
+            cout << "\033[01;32mInvalid input: [" << c << "]\033[0m" << endl;
```
[the remark this answers is not in the document]
Comment (rbx) [reply fairmq/plugins/DDS/runDDSCommandUI.cxx]: False positive.

File: fairmq/Transports.h [hunk 4]
```diff
@@ -53,7 +37,7 @@ static std::unordered_map<std::string, Transport> TransportTypes {
     { "ofi", Transport::OFI }
 };
 
-static std::unordered_map<Transport, std::string> TransportNames {
+static std::unordered_map<Transport, std::string, tools::HashEnum<Transport>> TransportNames {
```
Comment (dennisklein) [inline fairmq/Transports.h]: Pls revert this as discussed.

File: examples/dds/fairmq-start-ex-dds.sh.in [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+################################################################################
+
+set -e
+
+cleanup() {
+  dds-session stop $1
+  echo "CLEANUP PERFORMED"
+}
+
+source @DDS_INCDIR@/../DDS_env.sh
```
Comment (dennisklein) [inline examples/dds/fairmq-start-ex-dds.sh.in]:
```suggestion
source @DDS_INSTALL_PREFIX@/DDS_env.sh
```

**Comment by dennisklein** [comment]
I rebased the whole thing again, there was some strange double commit 9dc1fdb58f1f02407d506955bd145d6cfb51f2c7 which duplicated a commit from Oct11, 2018 in here.

**Comment by dennisklein** [comment]
@rbx Still cannot push to your branch, have you checked the box in the sidebar here?

Find my rebase and additions meanwhile in https://github.com/dennisklein/FairMQ/tree/rbx/dennis_wait_mode2.

File: examples/dds/fairmq-start-ex-dds.sh.in (hunk 5 above)
Comment (dennisklein) [reply examples/dds/fairmq-start-ex-dds.sh.in]: This is already included in my rebased branch.

**Comment by dennisklein** [comment]
Relative executable path + ssh plugin is currently blocked by FairRootGroup/DDS#229. Let's wait for an answer first, if there is no quick resolution, I will revert back to absolute path for now :(

**Comment by rbx** [comment]
Checkbox is on since yesterday. I pushed your branch here.

**Pull request merged**